core/test/utils/assertions.hpp

```diff
@@ -127,15 +132,17 @@ double get_relative_error(const MatrixData1 &first, const MatrixData2 &second)
                 get_next_value(first_it, end(first.nonzeros), row, col);
             const auto second_val =
                 get_next_value(second_it, end(second.nonzeros), row, col);
-            diff += squared_norm(first_val - second_val);
-            first_norm += squared_norm(first_val);
-            second_norm += squared_norm(second_val);
+            const double first = abs(first_val);
+            const double second = abs(second_val);
+            diff += squared_norm(first - second);
+            first_norm += squared_norm(first);
+            second_norm += squared_norm(second);
```

Comment:
Here we are only comparing the element magnitudes
```suggestion
            diff += squared_norm(first_val - second_val);
            first_norm += squared_norm(first_val);
            second_norm += squared_norm(second_val);
```

Reply:
Right, now that I changed it back I realize why I was doing it like this... See the compiling errors. I'll try to come up with something better nonetheless.

Reply:
Did you try using `std::max<double>` already?

Reply:
I fixed this with a templated struct which builds the proper type from two real types. It's also tested in assertion_test.